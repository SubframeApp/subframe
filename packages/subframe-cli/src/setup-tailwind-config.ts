import { join } from "node:path"
import ora from "ora"
import prompts from "prompts"
import { ObjectLiteralExpression, printNode, Project, QuoteKind, SourceFile, SyntaxKind } from "ts-morph"
import { abortOnState } from "./sync-helpers"
import { makeSubframeContentGlob, makeSubframeRequire } from "./transforms/tailwind"

function printManualTailwindSteps(cwd: string, subframeDirPath: string, prependText: string) {
  const subframePresetRequire = printNode(makeSubframeRequire(cwd, subframeDirPath))
  const subframeContentGlob = makeSubframeContentGlob(cwd, subframeDirPath)

  const warningMessage = `${prependText}
1. Append the following to the presets array, or create if it doesn't exist: ${subframePresetRequire}
2. Append the following to the content array: ${subframeContentGlob}

Your end results should look something like this:

module.exports = {
  presets: [${subframePresetRequire}], // added by Subframe
  content: [
    // already existing content
    "${subframeContentGlob}", // added by Subframe
  ],
  // everything else
};
`

  // makes the warning message cyan via https://stackoverflow.com/questions/9781218/how-to-change-node-jss-console-font-color
  console.log("\x1b[36m%s\x1b[0m", warningMessage)
}
function _getQuoteChar(configObject: ObjectLiteralExpression) {
  return configObject.getFirstDescendantByKind(SyntaxKind.StringLiteral)?.getQuoteKind() === QuoteKind.Single
    ? "'"
    : '"'
}

export async function setupTailwindConfig(cwd: string, subframeDirPath: string, opts: { tailwind?: boolean }) {
  const subframePresetRequireAST = makeSubframeRequire(cwd, subframeDirPath)
  const subframeContentGlob = makeSubframeContentGlob(cwd, subframeDirPath)

  const project = new Project({ compilerOptions: { allowJs: true } })

  project.addSourceFileAtPathIfExists(join(cwd, "tailwind.config.ts"))
  project.addSourceFileAtPathIfExists(join(cwd, "tailwind.config.js"))

  const tailwindConfig =
    project.getSourceFile(join(cwd, "tailwind.config.ts")) ?? project.getSourceFile(join(cwd, "tailwind.config.js"))

  /** no Tailwind config, let's skip this step then */
  if (!tailwindConfig) {
    printManualTailwindSteps(
      cwd,
      subframeDirPath,
      "Subframe could not find a tailwind.config.js file. If you use a .ts file, you'll need to configure it manually:",
    )
    return
  }

  /** config before transformations */
  const initialText = tailwindConfig.print()

  // na√Øve heuristics to check that the
  // subframe preset and content glob have
  // been properly set up
  if (
    initialText.indexOf(
      // example: require("./<path-to-subframe>/tailwind.config.js")
      printNode(subframePresetRequireAST),
    ) !== -1 &&
    initialText.indexOf(
      // example: "./<path-to-subframe>/**/*.{tsx,ts,js,jsx}"
      subframeContentGlob,
    ) !== -1
  ) {
    return
  }

  prompts.override({
    updateTailwindConfig: opts.tailwind,
  })
  const response = await prompts({
    type: "confirm",
    name: "updateTailwindConfig",
    initial: true,
    message: "Do you want Subframe to configure your Tailwind config?",
    onState: abortOnState,
  })
  if (!response.updateTailwindConfig) {
    return
  }

  const spinner = ora("Updating Tailwind config").start()
  transformTailwindConfigFile(tailwindConfig, cwd, subframeDirPath)
  await tailwindConfig.save()
  spinner.succeed("Tailwind config updated")
}

// helper function that's exported for testing
// NOTE: mutates tailwindConfig file in place
export function transformTailwindConfigFile(tailwindConfig: SourceFile, cwd: string, subframeDirPath: string) {
  const initialText = tailwindConfig.print()

  const configObject = tailwindConfig.getDescendantsOfKind(SyntaxKind.ObjectLiteralExpression)[0]
  if (!configObject) {
    printManualTailwindSteps(
      cwd,
      subframeDirPath,
      "Subframe could not find a valid Tailwind config object. To setup manually:",
    )
    return
  }

  const subframePresetRequireAST = makeSubframeRequire(cwd, subframeDirPath)
  const subframeContentGlob = makeSubframeContentGlob(cwd, subframeDirPath)

  const subframePresetAstText = printNode(subframePresetRequireAST)

  addTailwindConfigContent(configObject, subframeContentGlob)
  addTailwindConfigPresets(configObject, subframePresetAstText)

  /** config after transformations */
  const finalText = tailwindConfig.print()

  /**
   * No material changes could be made to the config,
   * so we warn the user and skip this step.
   *
   * This could be due to:
   * - the config structure being different than expected
   * - the config already being set up
   * - the config being empty
   * - the config file being invalid
   */
  if (initialText === finalText) {
    printManualTailwindSteps(
      cwd,
      subframeDirPath,
      "Subframe could not automatically configure your Tailwind config. To setup manually:",
    )
    return
  }

  tailwindConfig.formatText({
    indentSize: 2,
  })
}

// Yoinked from: https://github.com/shadcn-ui/ui/blob/main/packages/shadcn/src/utils/updaters/update-tailwind-content.ts#L75
async function addTailwindConfigContent(configObject: ObjectLiteralExpression, content: string) {
  const quoteChar = _getQuoteChar(configObject)
  const existingProperty = configObject
    .getDescendantsOfKind(SyntaxKind.PropertyAssignment)
    .find((property) => property.getName().includes("content"))

  if (!existingProperty) {
    const newProperty = {
      name: "content",
      initializer: `[${quoteChar}${content}${quoteChar}]`,
    }
    configObject.addPropertyAssignment(newProperty)

    return configObject
  }

  if (existingProperty.isKind(SyntaxKind.PropertyAssignment)) {
    const initializer = existingProperty.getInitializer()

    // If property is an array, append.
    if (initializer?.isKind(SyntaxKind.ArrayLiteralExpression)) {
      const newValue = `${quoteChar}${content}${quoteChar}`

      // Check if the array already contains the value.
      if (
        initializer
          .getElements()
          .map((element) => element.getText())
          .includes(newValue)
      ) {
        return configObject
      }

      initializer.addElement(newValue)
    }

    return configObject
  }

  return configObject
}

async function addTailwindConfigPresets(configObject: ObjectLiteralExpression, presets: string) {
  const existingProperty = configObject
    .getDescendantsOfKind(SyntaxKind.PropertyAssignment)
    .find((property) => property.getName().includes("presets"))

  if (!existingProperty) {
    const newProperty = {
      name: "presets",
      initializer: `[${presets}]`,
    }
    configObject.addPropertyAssignment(newProperty)

    return configObject
  }

  if (existingProperty.isKind(SyntaxKind.PropertyAssignment)) {
    const initializer = existingProperty.getInitializer()

    // If property is an array, append.
    if (initializer?.isKind(SyntaxKind.ArrayLiteralExpression)) {
      const newValue = presets

      // Check if the array already contains the value.
      if (
        initializer
          .getElements()
          .map((element) => element.getText())
          .includes(newValue)
      ) {
        return configObject
      }

      initializer.addElement(newValue)
    }

    return configObject
  }

  return configObject
}
