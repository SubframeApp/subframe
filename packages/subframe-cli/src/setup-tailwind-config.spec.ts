import { transformTailwindConfigContent } from "./setup-tailwind-config"

// Taken from Next13's default tailwind config
const DEFAULT_TAILWIND_CONFIG = {
  content: [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
  ],
  theme: {
    extend: {
      backgroundImage: {
        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
        "gradient-conic": "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))",
      },
    },
  },
  plugins: [],
}

// Taken from Tailwind's Vite guide
// Modified to have a presets property
// https://tailwindcss.com/docs/guides/vite
const VITE_TAILWIND_CONFIG_WITH_PRESETS_PROPERTY = {
  content: ["./index.html", "./src/**/*.{js,ts,jsx,tsx}"],
  theme: {
    extend: {},
  },
  plugins: [],
  presets: [],
}

const CWD = "cwd/path"
const SUBFRAME_DIR_PATH = "subframe-dir/path"

function makeTailwindConfigContents(config: object, type: "esm" | "cjs" | "ts" = "cjs") {
  switch (type) {
    case "esm":
      return `/** @type {import('tailwindcss').Config} */
export default ${JSON.stringify(config, null, 2)} satisfies Config`
    case "cjs":
      return `/** @type {import('tailwindcss').Config} */
  module.exports = ${JSON.stringify(config, null, 2)}
  `
    case "ts":
      return `import type { Config } from 'tailwindcss'
      export default ${JSON.stringify(config, null, 2)} satisfies Config`
  }
}

async function getTransformedTailwindConfigFile(
  config: object,
  type: "esm" | "cjs" | "ts" = "cjs",
): Promise<[string, string]> {
  // Taken from Next13's default tailwind config
  const contents = makeTailwindConfigContents(config, type)
  const before = contents

  const after = await transformTailwindConfigContent(
    before,
    SUBFRAME_DIR_PATH,
    `${CWD}/tailwind.config.${type === "ts" ? "ts" : "js"}`,
  )
  return [before, after]
}

describe("#transformTailwindConfig", () => {
  it("adds content to empty NextJS tailwind config", async () => {
    const [_before, after] = await getTransformedTailwindConfigFile({})
    expect(after).toMatchSnapshot()
  })

  it("adds content to default NextJS tailwind config", async () => {
    const [_before, after] = await getTransformedTailwindConfigFile(DEFAULT_TAILWIND_CONFIG)
    expect(after).toMatchSnapshot()
  })

  it("adds content to default Vite tailwind config with presets property", async () => {
    const [_before, after] = await getTransformedTailwindConfigFile(VITE_TAILWIND_CONFIG_WITH_PRESETS_PROPERTY)
    expect(after).toMatchSnapshot()
  })

  it("adds content to ESM-style file", async () => {
    const [_before, after] = await getTransformedTailwindConfigFile(DEFAULT_TAILWIND_CONFIG, "esm")
    expect(after).toMatchSnapshot()
  })

  it("adds content to TypeScript file", async () => {
    const [_before, after] = await getTransformedTailwindConfigFile(DEFAULT_TAILWIND_CONFIG, "ts")
    expect(after).toMatchSnapshot()
  })

  it("running it the second time does not change the file", async () => {
    let [_before, after] = await getTransformedTailwindConfigFile(DEFAULT_TAILWIND_CONFIG, "ts")
    const firstRun = after
    after = await transformTailwindConfigContent(after, SUBFRAME_DIR_PATH, `${CWD}/tailwind.config.ts`)
    expect(after).toEqual(firstRun)
  })
})
