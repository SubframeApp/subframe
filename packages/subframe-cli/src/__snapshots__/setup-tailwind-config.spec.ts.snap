// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#transformTailwindConfig adds content to ESM-style file 1`] = `
"/** @type {import('tailwindcss').Config} */
export default {
  "content": [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
        "./subframe-dir/path/**/*.{tsx,ts,js,jsx}"
    ],
  "theme": {
    "extend": {
      "backgroundImage": {
        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
        "gradient-conic": "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))"
      }
    }
  },
  "plugins": [],
    presets: [(await import("./subframe-dir/path/tailwind.config")).default]
} satisfies Config"
`;

exports[`#transformTailwindConfig adds content to TypeScript file 1`] = `
"import type { Config } from 'tailwindcss'
      export default {
  "content": [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
              "./subframe-dir/path/**/*.{tsx,ts,js,jsx}"
        ],
  "theme": {
    "extend": {
      "backgroundImage": {
        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
        "gradient-conic": "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))"
      }
    }
  },
  "plugins": [],
          presets: [(await import("./subframe-dir/path/tailwind.config")).default as unknown as Partial<Config>]
    } satisfies Config"
`;

exports[`#transformTailwindConfig adds content to default NextJS tailwind config 1`] = `
"/** @type {import('tailwindcss').Config} */
  module.exports = {
  "content": [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
          "./subframe-dir/path/**/*.{tsx,ts,js,jsx}"
    ],
  "theme": {
    "extend": {
      "backgroundImage": {
        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
        "gradient-conic": "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))"
      }
    }
  },
  "plugins": [],
      presets: [(await import("./subframe-dir/path/tailwind.config")).default]
}
  "
`;

exports[`#transformTailwindConfig adds content to default Vite tailwind config with presets property 1`] = `
"/** @type {import('tailwindcss').Config} */
  module.exports = {
  "content": [
    "./index.html",
    "./src/**/*.{js,ts,jsx,tsx}",
          "./subframe-dir/path/**/*.{tsx,ts,js,jsx}"
    ],
  "theme": {
    "extend": {}
  },
  "plugins": [],
  "presets": [(await import("./subframe-dir/path/tailwind.config")).default]
}
  "
`;

exports[`#transformTailwindConfig adds content to empty NextJS tailwind config 1`] = `
"/** @type {import('tailwindcss').Config} */
  module.exports = {}
  "
`;
